sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dofn/DetectNewPartitionsDoFn.java

```diff
@@ -109,9 +118,15 @@ public TimestampRange initialRestriction(@Element PartitionMetadata partition) {
         TimestampUtils.previous(createdAt), com.google.cloud.Timestamp.MAX_VALUE);
   }
 
+  @GetSize
+  public double getSize() {
+    double throughput = throughputEstimator.get();
+    LOG.debug("Reported getSize() - throughput: " + throughput);
+    return throughput;
```

Comment:
This might be confusing as we report `timeGap * throughput` in ReadChangeStreamPartitionDoFn: 

https://github.com/apache/beam/blob/e0e10b9e5432643b884c381d145e5924cc4ef193/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/dofn/ReadChangeStreamPartitionDoFn.java#L149-L166

But here, we only report the throughput. 

Just a bold idea. Can we totally rely on ReadChangeStreamPartitionDoFn for autoscaling? In other words, we always return 0 for getSize() in DetectNewPartitionsDoFn.

Reply:
I initially had the `timeGap * throughput`, but I don't think it adds any value here. In fact, the backlog bytes increases as time passes, because the `timeGap` becomes bigger until there is a partition to schedule. This makes it very weird in terms of monitoring as it gives the false sense that the backlog is increasing were it could be constant.

I am a bit worried about always returning a `0` here, as it could have unintended consequences (not sure which). I will consult with other team members and see if it makes sense.

Reply:
Reporting zero seems good if you never want to upscale/downscale the pipeline based upon this stage.  Dataflow autoscaling is currently based upon max of needed stages considering each stage independently.
Would this DoFn ever require to run on more than 1 worker? Would it ever be resource constrained?

Another idea would be to have some threshold for which you report zero instead of the calculated backlog. This would help not trigger spurious upscaling due to the backlog calculation when things are keeping up but could give a signal in the cases things were behind.

I also agree reporting recent throughput as backlog is confusing, it seems like this will always be non-zero?  Is there a way to tell if timeGap is increasing because there is nothing to do or it is actually falling behind? That seems like a prerequisite for reporting backlog correctly.

Reply:
> Would this DoFn ever require to run on more than 1 worker? Would it ever be resource constrained?

Not with the current design, it works as singleton scheduler at the moment.

> Another idea would be to have some threshold for which you report zero instead of the calculated backlog.

We thought about this, I will try to experiment with certain thresholds.

> I also agree reporting recent throughput as backlog is confusing, it seems like this will always be non-zero?

Yes, the SDF only advances the restriction (`tryClaim`) when it successfully outputs a partition. We need to do this, since we use the restriction start time to fetch the next batch of partitions that were created after that time. This also helps with replay protection, meaning that even we schedule a partition (update the database state), but are unable to claim it, on the next try we would get it. There are possibly long periods that no partitions are scheduled, so the time gap would increase.

>  Is there a way to tell if timeGap is increasing because there is nothing to do or it is actually falling behind?

There is a way to do this which is to count the number of partitions to schedule given from the restriction start time. However, this would require us to call the database on every `GetSize` call, which I think could be expensive. I will try experimenting with this though.